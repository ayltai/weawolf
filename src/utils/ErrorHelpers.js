export const ErrorHelpers = {};

ErrorHelpers.handle = error => console.error(error);
