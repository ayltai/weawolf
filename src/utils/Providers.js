export const Providers = {};

Providers.ACCU_WEATHER     = 1;
Providers.DARK_SKY         = 0;
Providers.OPEN_WEATHER_MAP = 3;
Providers.WEATHERBIT       = 2;

Providers.isValid = provider => provider >= 0 && provider <= 3;
