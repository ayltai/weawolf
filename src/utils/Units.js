import { NumberFormatHelpers, } from './NumberFormatHelpers';

export const Units = {};

Units.SI       = 'si';
Units.IMPERIAL = 'imperial';

Units.toTemperature = (temperature, unit, digits = 1, withSymbol = true) => unit === Units.SI ? `${NumberFormatHelpers.toFixed(temperature, digits)}${withSymbol ? '°C' : ''}` : `${NumberFormatHelpers.toFixed(temperature * 1.8 + 32, digits)}${withSymbol ? '°F' : ''}`;

Units.toWindSpeed = (windSpeed, unit, digits = 1, withSymbol = true) => unit === Units.SI ? `${NumberFormatHelpers.toFixed(windSpeed, digits)}${withSymbol ? 'km/h' : ''}` : `${NumberFormatHelpers.toFixed(windSpeed / 1.60934, digits)}${withSymbol ? 'mi/h' : ''}`;
