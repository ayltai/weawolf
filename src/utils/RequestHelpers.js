import superagent from 'superagent';

import { ErrorHelpers, } from './ErrorHelpers';

export const RequestHelpers = {};

RequestHelpers.request = async (url, argsValidator = () => {}) => {
    argsValidator();

    const response = await superagent.get(url).set('Accept', 'application/json');

    if (response.status >= 200 && response.status < 300) return response.body;

    ErrorHelpers.handle(new Error(`Received HTTP ${response.status} for URL ${url}`));
};
