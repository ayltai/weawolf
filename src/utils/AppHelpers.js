import i18next from 'i18next';
import moment from 'moment';

import { Configurations, } from '../Configurations';
import { RequestHelpers, } from './RequestHelpers';

export const AppHelpers = {};

AppHelpers.setAutoLaunch = enabled => {
    const app = window.require('electron').remote.app;

    app.setLoginItemSettings({
        openAtLogin  : enabled,
        openAsHidden : true,
        path         : app.getPath('exe'),
    });
};

AppHelpers.changeLocale = locale => {
    i18next.changeLanguage(locale);
    moment.locale(locale);
};

AppHelpers.checkForUpdates = async () => {
    const response = RequestHelpers.request(Configurations.UPDATE_URL);

    const extension = process.platform === 'darwin' ? 'dmg' : process.platform === 'win32' ? 'exe' : 'AppImage';
    return [ Configurations.APP_VERSION !== response.version, `https://github.com/ayltai/weawolf/archive/release/Weawolf-${response.version}.${process.platform}.${extension}`, ];
};
