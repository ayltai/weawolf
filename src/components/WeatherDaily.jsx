import { Grid, Tooltip, } from '@material-ui/core';
import moment from 'moment';
import PropTypes from 'prop-types';
import React, { Fragment, } from 'react';
import { useTranslation, } from 'react-i18next';

import { NumberFormatHelpers, } from '../utils/NumberFormatHelpers';
import { Providers, } from '../utils/Providers';
import { Units, } from '../utils/Units';
import { WeatherHelpers, } from '../utils/WeatherHelpers';
import { Label, } from './Label';
import { Temperature, } from './Temperature';

export const WeatherDaily = props => {
    const { t, } = useTranslation();

    const summary = props.summary.charAt(0).toUpperCase() + props.summary.slice(1);

    let i = 0;

    const date = (
        <Label
            key={i++}
            align='center'
            variant='body2'
            noShadow={props.noShadow}
            noWrap>
            {moment(props.date || new Date()).format('ddd')}
        </Label>
    );

    const icon = (
        <Label
            key={i++}
            align='center'
            variant='h4'
            noShadow={props.noShadow}
            noWrap>
            <i className={`wi ${WeatherHelpers.toIcon(props.provider, props.iconId)}`}></i>
        </Label>
    );

    const temperatureHigh = (
        <Temperature
            key={i++}
            align='center'
            variant='body2'
            unit={props.unit}
            prefix={t(props.temperatureHighPrefix)}
            noShadow={props.noShadow}
            noWrap>
            {props.temperatureHigh}
        </Temperature>
    );

    const temperatureLow = (
        <Temperature
            key={i++}
            align='center'
            variant='body2'
            unit={props.unit}
            prefix={t(props.temperatureLowPrefix)}
            noShadow={props.noShadow}
            noWrap>
            {props.temperatureLow}
        </Temperature>
    );

    const precip = (
        <Label
            key={i++}
            align='center'
            variant='body2'
            noShadow={props.noShadow}
            noWrap>
            {`${t(props.precipPrefix || '')}${Math.round(props.precipProbability * 100)}%`}
        </Label>
    );

    const blocks = [];

    if (props.orientation === 'horizontal') {
        blocks.push(
            <Grid
                key={i++}
                item
                xs={2}>
                {date}
                {precip}
            </Grid>
        );

        blocks.push(
            <Grid
                key={i++}
                item
                xs={2}>
                {icon}
            </Grid>
        );

        blocks.push(
            <Grid
                key={i++}
                item
                xs={6}>
                <Label
                    align='left'
                    variant='caption'
                    noShadow={props.noShadow}
                    noWrap>
                    {summary}
                </Label>
            </Grid>
        );

        blocks.push(
            <Grid
                key={i}
                item
                xs={2}>
                {(props.temperatureHighPrefix || 'ðŸ¥µ') + ' ' + temperatureHigh}
                {(props.temperatureLowPrefix || 'ðŸ¥¶') + ' ' + temperatureLow}
            </Grid>
        );
    } else if (props.orientation === 'vertical') {
        blocks.push(date);
        blocks.push(icon);
        blocks.push(temperatureHigh);
        blocks.push(temperatureLow);
        blocks.push(precip);
    }

    return (
        <Tooltip title={
            <Fragment>
                {moment(props.date).format('LL')}<br />
                {summary}<br />
                {`${t(props.temperaturePrefix || '')}${Units.toTemperature(props.temperatureLow, props.unit)} - ${Units.toTemperature(props.temperatureHigh, props.unit)}`}<br />
                {`${t(props.precipPrefix || '')}${NumberFormatHelpers.toFixed(props.precipProbability * 100, 0)}% ${NumberFormatHelpers.toFixed(props.precipIntensity)}mm`}<br />
                {props.humidity ? (
                    <>
                        {`${t(props.humidityPrefix || '')}${NumberFormatHelpers.toFixed(props.humidity * 100, 0)}%`}<br />
                    </>
                ) : null}
                {`${t(props.windSpeedPrefix || '')}${Units.toWindSpeed(props.windSpeed, props.unit, 0)}`}<br />
                {`${t(props.uvIndexPrefix || '')}${NumberFormatHelpers.toFixed(props.uvIndex, 0)}`}
            </Fragment>
        }>
            <Grid
                container
                direction={props.orientation === 'vertical' ? 'column' : 'row'}
                alignItems='center'
                justify='space-evenly'>
                {blocks}
            </Grid>
        </Tooltip>
    );
};

WeatherDaily.propTypes = {
    orientation           : PropTypes.oneOf([
        'horizontal',
        'vertical',
    ]),
    date                  : PropTypes.object,
    provider              : PropTypes.oneOf([
        Providers.ACCU_WEATHER,
        Providers.DARK_SKY,
        Providers.OPEN_WEATHER_MAP,
        Providers.WEATHERBIT,
    ]),
    unit                  : PropTypes.oneOf([
        Units.SI,
        Units.IMPERIAL,
    ]),
    summary               : PropTypes.string,
    iconId                : PropTypes.oneOfType([
        PropTypes.number,
        PropTypes.string,
    ]),
    temperaturePrefix     : PropTypes.string,
    temperatureHighPrefix : PropTypes.string,
    temperatureHigh       : PropTypes.number,
    temperatureLowPrefix  : PropTypes.string,
    temperatureLow        : PropTypes.number,
    humidityPrefix        : PropTypes.string,
    humidity              : PropTypes.number,
    precipPrefix          : PropTypes.string,
    precipProbability     : PropTypes.number,
    precipIntensity       : PropTypes.number,
    windSpeedPrefix       : PropTypes.string,
    windSpeed             : PropTypes.number,
    uvIndexPrefix         : PropTypes.string,
    uvIndex               : PropTypes.number,
    noShadow              : PropTypes.bool,
};
