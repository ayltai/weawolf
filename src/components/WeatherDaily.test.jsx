import { mount, } from 'enzyme';
import React from 'react';
import renderer from 'react-test-renderer';

import { Providers, } from '../utils/Providers';
import { Units, } from '../utils/Units';
import { WeatherDaily, } from './WeatherDaily';

jest.mock('react-i18next', () => ({
    useTranslation : () => ({
        t : key => key,
    }),
}));

const date              = new Date(1587628337);
const summary           = 'Partly Cloudy';
const iconId            = 'partly-cloudy-night';
const provider          = Providers.DARK_SKY;
const unit              = Units.SI;
const temperatureHigh   = 29.7;
const temperatureLow    = 22.4;
const humidity          = 0.77;
const precipProbability = 0.48;
const precipIntensity   = 0.2;
const windSpeed         = 7;
const uvIndex           = 4;

const componentHorizontal = (
    <WeatherDaily
        orientation='horizontal'
        date={date}
        summary={summary}
        provider={provider}
        unit={unit}
        iconId={iconId}
        temperatureHigh={temperatureHigh}
        temperatureLow={temperatureLow}
        humidity={humidity}
        precipProbability={precipProbability}
        precipIntensity={precipIntensity}
        windSpeed={windSpeed}
        uvIndex={uvIndex} />
);

const componentVertical = (
    <WeatherDaily
        orientation='vertical'
        date={date}
        summary={summary}
        provider={provider}
        unit={unit}
        iconId={iconId}
        temperatureHigh={temperatureHigh}
        temperatureLow={temperatureLow}
        humidity={humidity}
        precipProbability={precipProbability}
        precipIntensity={precipIntensity}
        windSpeed={windSpeed}
        uvIndex={uvIndex} />
);

describe('<WeatherDaily />', () => {
    it('renders correctly', () => {
        expect(renderer.create(componentHorizontal).toJSON()).toMatchSnapshot();
    });

    it('renders correctly', () => {
        expect(renderer.create(componentVertical).toJSON()).toMatchSnapshot();
    });

    it('mounts without errors', () => {
        mount(componentHorizontal);
    });

    it('mounts without errors', () => {
        mount(componentVertical);
    });
});
