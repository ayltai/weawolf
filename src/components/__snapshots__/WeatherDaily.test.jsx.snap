// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<WeatherDaily /> renders correctly 1`] = `
<div
  aria-describedby={null}
  className="MuiGrid-root MuiGrid-container MuiGrid-align-items-xs-center MuiGrid-justify-xs-space-evenly"
  onBlur={[Function]}
  onFocus={[Function]}
  onMouseLeave={[Function]}
  onMouseOver={[Function]}
  onTouchEnd={[Function]}
  onTouchStart={[Function]}
  title={null}
>
  <div
    className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-2"
  >
    <p
      aria-describedby={null}
      className="MuiTypography-root makeStyles-shadow-2 MuiTypography-body2 MuiTypography-noWrap MuiTypography-alignCenter"
      onBlur={[Function]}
      onFocus={[Function]}
      onMouseLeave={[Function]}
      onMouseOver={[Function]}
      onTouchEnd={[Function]}
      onTouchStart={[Function]}
      title=""
    >
      Mon
    </p>
    <p
      aria-describedby={null}
      className="MuiTypography-root makeStyles-shadow-4 MuiTypography-body2 MuiTypography-noWrap MuiTypography-alignCenter"
      onBlur={[Function]}
      onFocus={[Function]}
      onMouseLeave={[Function]}
      onMouseOver={[Function]}
      onTouchEnd={[Function]}
      onTouchStart={[Function]}
      title=""
    >
      48%
    </p>
  </div>
  <div
    className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-2"
  >
    <p
      aria-describedby={null}
      className="MuiTypography-root makeStyles-shadow-6 MuiTypography-h4 MuiTypography-noWrap MuiTypography-alignCenter"
      onBlur={[Function]}
      onFocus={[Function]}
      onMouseLeave={[Function]}
      onMouseOver={[Function]}
      onTouchEnd={[Function]}
      onTouchStart={[Function]}
      title=""
    >
      <i
        className="wi wi-night-alt-cloudy"
      />
    </p>
  </div>
  <div
    className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6"
  >
    <p
      aria-describedby={null}
      className="MuiTypography-root makeStyles-shadow-8 MuiTypography-caption MuiTypography-noWrap MuiTypography-alignLeft"
      onBlur={[Function]}
      onFocus={[Function]}
      onMouseLeave={[Function]}
      onMouseOver={[Function]}
      onTouchEnd={[Function]}
      onTouchStart={[Function]}
      title=""
    >
      Partly Cloudy
    </p>
  </div>
  <div
    className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-2"
  >
    ðŸ¥µ [object Object]
    ðŸ¥¶ [object Object]
  </div>
</div>
`;

exports[`<WeatherDaily /> renders correctly 2`] = `
<div
  aria-describedby={null}
  className="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-column MuiGrid-align-items-xs-center MuiGrid-justify-xs-space-evenly"
  onBlur={[Function]}
  onFocus={[Function]}
  onMouseLeave={[Function]}
  onMouseOver={[Function]}
  onTouchEnd={[Function]}
  onTouchStart={[Function]}
  title={null}
>
  <p
    aria-describedby={null}
    className="MuiTypography-root makeStyles-shadow-10 MuiTypography-body2 MuiTypography-noWrap MuiTypography-alignCenter"
    onBlur={[Function]}
    onFocus={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onTouchEnd={[Function]}
    onTouchStart={[Function]}
    title=""
  >
    Mon
  </p>
  <p
    aria-describedby={null}
    className="MuiTypography-root makeStyles-shadow-12 MuiTypography-h4 MuiTypography-noWrap MuiTypography-alignCenter"
    onBlur={[Function]}
    onFocus={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onTouchEnd={[Function]}
    onTouchStart={[Function]}
    title=""
  >
    <i
      className="wi wi-night-alt-cloudy"
    />
  </p>
  <p
    aria-describedby={null}
    className="MuiTypography-root makeStyles-shadow-14 MuiTypography-body2 MuiTypography-noWrap MuiTypography-alignCenter"
    onBlur={[Function]}
    onFocus={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onTouchEnd={[Function]}
    onTouchStart={[Function]}
    title="29.7Â°C"
  >
    30Â°C
  </p>
  <p
    aria-describedby={null}
    className="MuiTypography-root makeStyles-shadow-16 MuiTypography-body2 MuiTypography-noWrap MuiTypography-alignCenter"
    onBlur={[Function]}
    onFocus={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onTouchEnd={[Function]}
    onTouchStart={[Function]}
    title="22.4Â°C"
  >
    22Â°C
  </p>
  <p
    aria-describedby={null}
    className="MuiTypography-root makeStyles-shadow-18 MuiTypography-body2 MuiTypography-noWrap MuiTypography-alignCenter"
    onBlur={[Function]}
    onFocus={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onTouchEnd={[Function]}
    onTouchStart={[Function]}
    title=""
  >
    48%
  </p>
</div>
`;
